### PR TITLE
C4-485 Use batchId = 1 for EOA flow, batchId = 0 for AA flow

### PR DESCRIPTION
**Pull request by filmakarov**
Removed the support of 2d nonce from SCW implementation.
Kept the mapping of 2d nonces in storage
Aways accessing batchId 0 in the code where ever nonce is needed.

**Open Question:** actions steps decided still not solve the [issue #246](https://github.com/code-423n4/2023-01-biconomy-findings/issues/246). As conflicts between AA/EP flow and execTransaction flows are present even for 1d nonces.
However, I think that is not a big problem now, as there are not many AA transaction by now. So there won't be a lot of conflicts.
Should initiate a discussion with 4337 community regarding that.

* The quick fix could be using `nonces[1]` for execTransaction flow. 

Can quickly update if we decide to use this.

I decided not to remove `batchId` from  `execTransaction` parameters for future compatibility
Updated NatSpec to reflect this.

Difference between using uint256 vs mapping to access nonce is 27 gas units.
So for 1,000 txns it will be about $.01 on Polygon or $10 for 1M txns. So makes sense to still use mapping.

### CODE REVIEW

**Review by livingrockrises — COMMENTED** [review]
don't upload yarn.lock and package-lock.json file in intermediate PRs for now

**Comment by livingrockrises** [comment]
momentarily (maybe temp child branch) change 2D to 1D everywhere and compare gas results at transactional layer + contract deployment cost.

**Comment by filmakarov** [comment]
> momentarily (maybe temp child branch) change 2D to 1D everywhere and compare gas results at transactional layer + contract deployment cost.

2d to 1d = do you mean change mapping to uint256 ?

transactional layer = do you mean runtime gas costs? 

> Difference between using uint256 vs mapping to access nonce is 27 gas units.
> So for 1,000 txns it will be about $.01 on Polygon or $10 for 1M txns. So makes sense to still use mapping.

Does it answer the question regarding runtime costs?

**Comment by livingrockrises** [comment]
> > momentarily (maybe temp child branch) change 2D to 1D everywhere and compare gas results at transactional layer + contract deployment cost.
> 
> 2d to 1d = do you mean change mapping to uint256 ?
> 
> transactional layer = do you mean runtime gas costs?
> 
> > Difference between using uint256 vs mapping to access nonce is 27 gas units.
> > So for 1,000 txns it will be about $.01 on Polygon or $10 for 1M txns. So makes sense to still use mapping.
> 
> Does it answer the question regarding runtime costs?
> 
> =============
> 
> 2. I added three tests regarding nonces to the `testGroup1.ts`. Is it ok by now, or should I move them to the separate file? In any case, we definitely need refactoring for this file. Should I plan it for the next week?

1. Yes what's the overhead if we leave a room for 2D by keeping 2D state variable compared to just uint256.
2. Hmm test cases definitely need some refactoring at the files level. Can plan for it next week as we approach overall test coverage and writing more test cases.

**Comment by filmakarov** [comment]
> 1. Yes what's the overhead if we leave a room for 2D by keeping 2D state variable compared to just uint256.

**deployCounterFactualWallet**
```
2d: 261108
1d: 261086
```

**deploy Smart Account Singleton**
```
2d: 3,362,397
1d: 3,321,574

```

**execTransaction**
```
2d: 207666
1d: 207614
```

**Comment by livingrockrises** [comment]
we can go ahead with locking nonces[1] for execTransaction flow. instead of nonces[0]

**Comment by filmakarov** [comment]
@livingrockrises check please

**Comment by livingrockrises** [comment]
> @livingrockrises check please

checking

File: contracts/smart-contract-wallet/MaliciousAccount.sol [hunk 1]
```diff
@@ -208,16 +208,13 @@ contract MaliciousAccount is
                     // Payment info
                     refundInfo,
                     // Signature info
-                    nonces[batchId]
+                    nonces[0]++
```
Comment (livingrockrises) [inline contracts/smart-contract-wallet/MaliciousAccount.sol:211]: should this be 1 as well if changes made elsewhere

File: contracts/smart-contract-wallet/SmartAccountNoAuth.sol [hunk 2]
```diff
@@ -219,11 +219,9 @@ contract SmartAccountNoAuth is
                     // Payment info
                     refundInfo,
                     // Signature info
-                    nonces[batchId]
+                    nonces[0]++
```
Comment (livingrockrises) [inline contracts/smart-contract-wallet/SmartAccountNoAuth.sol:222]: this should be 1 as well.

File: test/smart-wallet/gasEstimations.ts [hunk 3]
```diff
@@ -470,7 +472,7 @@ describe("Wallet tx gas estimations with and without refunds", function () {
       to: token.address,
       // value: ethers.utils.parseEther("1"),
       data: encodeTransfer(charlie, ethers.utils.parseEther("10").toString()),
-      nonce: await userSCW.getNonce(0),
+      nonce: await userSCW.getNonce(EOA_CONTROLLED_FLOW),
```
Comment (livingrockrises) [inline test/smart-wallet/gasEstimations.ts:475]: I see now. should we keep this getNonce method?

Comment (livingrockrises) [reply test/smart-wallet/gasEstimations.ts:475]: ok we will have to use getNonce(1) moving forward for 'forward' transaction payload

File: test/smart-wallet/gasEstimations.ts [hunk 4]
```diff
@@ -179,8 +181,8 @@ describe("Wallet tx gas estimations with and without refunds", function () {
     const walletOwner = await userSCW.owner();
     expect(walletOwner).to.equal(owner);
 
-    const walletNonce1 = await userSCW.getNonce(0); // only 0 space is in the context now
-    const walletNonce2 = await userSCW.getNonce(1);
+    const walletNonce1 = await userSCW.getNonce(ACCOUNT_ABSTRACTION_FLOW);
```
Comment (livingrockrises) [inline test/smart-wallet/gasEstimations.ts:184]: maybe update the test in verifying-paymaster-test.ts to use 0 nonce space when creating and signing actual userop.

File: test/smart-wallet/gasEstimations.ts (hunk 3 above)
Comment (livingrockrises) [reply test/smart-wallet/gasEstimations.ts:475]: default nonce() method would return nonces[0] and getNonce(batchId) would return nonces[bacthId]

**Review by livingrockrises — COMMENTED** [review]
gave some final comments. looks good

**Comment by livingrockrises** [comment]
do resolve conflicts as well now. lmk if you need help

File: contracts/smart-contract-wallet/MaliciousAccount.sol (hunk 1 above)
Comment (filmakarov) [reply contracts/smart-contract-wallet/MaliciousAccount.sol:211]: fixed

File: contracts/smart-contract-wallet/SmartAccountNoAuth.sol (hunk 2 above)
Comment (filmakarov) [reply contracts/smart-contract-wallet/SmartAccountNoAuth.sol:222]: fixed

File: test/smart-wallet/gasEstimations.ts (hunk 3 above)
Comment (filmakarov) [reply test/smart-wallet/gasEstimations.ts:475]: yep

File: contracts/smart-contract-wallet/SmartAccount.sol [hunk 5]
```diff
@@ -204,7 +204,7 @@ contract SmartAccount is
     /// @dev Allows to execute a Safe transaction confirmed by required number of owners and then pays the account that submitted the transaction.
     /// Note: The fees are always transferred, even if the user transaction fails.
     /// @param _tx Wallet transaction 
-    /// @param batchId batchId key for 2D nonces
+    /// @param batchId batchId key for 2D nonces. Not used in this implementation. Left for future interface compatibility.
```
[the remark this answers is not in the document]
Comment (filmakarov) [reply contracts/smart-contract-wallet/SmartAccount.sol:198]: > we can write proper test cases for above exercise as part of writing test for our upgrade strategy.

yep, great idea

[the remark this answers is not in the document]
Comment (filmakarov) [reply contracts/smart-contract-wallet/SmartAccount.sol:198]: batchId removed

**Comment by filmakarov** [comment]
re-requested review @livingrockrises

File: test/utils/userOp.ts [hunk 6]
```diff
@@ -252,10 +253,10 @@ export async function fillUserOp(
       throw new Error("must have entryPoint to autofill nonce");
     const c = new Contract(
       op.sender!,
-      ["function nonce() view returns(address)"],
+      ["function nonce(uint256) view returns(uint256)"],
       provider
     );
-    op1.nonce = await c.nonce().catch(rethrow());
+    op1.nonce = await c.nonce(ACCOUNT_ABSTRACTION_FLOW).catch(rethrow());
```
Comment (livingrockrises) [inline test/utils/userOp.ts:259]: should we use nonce() method for AA flow [ which by default serves nonces[0] and name is also a standard ]
and getNonce method for forward transaction to get nonce for batchId 1 on client side.

and remove the overriding method nonce(_batchId) then it won't cause problems with ethers js as well

Comment (livingrockrises) [reply test/utils/userOp.ts:259]: @filmakarov

**Review by livingrockrises — APPROVED** [review]
approved

**Comment by livingrockrises** [comment]
@talhamalik883 please note that this change removes batchId from execTransaction function params. 
cc @tomarsachin2271

**Pull request merged**